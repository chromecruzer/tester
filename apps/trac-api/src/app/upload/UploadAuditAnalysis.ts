import {UploadAnalysisA} from "./UploadAnalysisA";
import {auditRecordFromJson, UploadAuditDetails} from "@trac/datatypes";
import util from "util";
import _ from "lodash";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class UploadAuditAnalysis extends UploadAnalysisA {
  public async comparisons(session) {
    return this.dataStore.retrieve(session)
      .then(auditRecord => auditRecordFromJson(auditRecord))
      .then(auditRecord => {
        // console.log(`retrieved audit record ${dump(auditRecord)}`);
        return auditRecord ? {
          location: auditRecord.location,
          matches: auditRecord.items.filter(i => i.audit_match === 'True Match'),
          moved: auditRecord.items.filter(i => i.audit_match === 'Found In Other Location'),
          nomatches: auditRecord.items.filter(i => i.audit_match === 'No Match'),
          missing: auditRecord.items.filter(i => i.audit_match === 'Missing'),
        } as UploadAuditDetails : null;
      })
      .catch(err => {
        throw new Error(`Session ${session} not found ${dump(err)}`);
      });
  }
}
