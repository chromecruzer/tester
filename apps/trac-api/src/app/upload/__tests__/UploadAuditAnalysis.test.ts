import UploadAuditAnalysis from "../UploadAuditAnalysis";

describe('UpdateAuditAnalysis', () => {
  let storageMock, auditRecord, analytics;
  beforeEach(() => {
    auditRecord = {
      location: {
        name: 'this is a location record',
        scan_date: new Date('10/10/2021'),
      },
      items: [
        {
          uuid: 'uuid1',
          audit_match: 'True Match',
        },
        {
          uuid: 'uuid2',
          audit_match: 'True Match'
        },
        {
          uuid: 'uuid3',
          audit_match: 'True Match'
        },
        {
          uuid: 'uuid4',
          audit_match: 'True Match'
        },
        {
          uuid: 'uuid5',
          audit_match: 'True Match'
        },
        {
          uuid: 'uuid6',
          audit_match: 'Found In Other Location'
        },
        {
          uuid: 'uuid7',
          audit_match: 'Found In Other Location'
        },
        {
          uuid: 'uuid8',
          audit_match: 'Found In Other Location'
        },
        {
          uuid: 'uuid9',
          audit_match: 'Expired'
        },
        {
          uuid: 'uuid10',
          audit_match: 'Expired'
        },
        {
          uuid: 'uuid11',
          audit_match: 'No Match'
        },
        {
          uuid: 'uuid12',
          audit_match: 'Missing'
        },
        {
          uuid: 'uuid13',
          audit_match: 'Missing'
        },
        {
          uuid: 'uuid14',
          audit_match: 'Missing'
        },
        {
          uuid: 'uuid15',
          audit_match: 'Missing'
        },
      ]
    };
    storageMock = {
      retrieve: jest.fn().mockReturnValue(Promise.resolve(auditRecord)),
    };
    analytics = new UploadAuditAnalysis(storageMock, 'schema name', 'table name')
  });
  describe('details', () => {
    it('should return the correct filtered results', () => {
      return analytics.details('I am a session')
        .then(actual => {
          expect(actual).toMatchSnapshot();
        });
    });
    it('should throw an error if the session is missing', () => {
    storageMock.retrieve.mockReturnValue(Promise.reject(new Error('Made to fail')));
    const fn = () => analytics.details('I am a session');
    expect(fn).rejects.toThrowErrorMatchingSnapshot();
    });
  });
  describe('analytics', () => {
    it('should return the correct analytics', () => {
      return analytics.analytics('I am a session')
        .then(actual => {
          expect(actual).toHaveProperty('matches', 5);
          expect(actual).toHaveProperty('moved', 3);
          expect(actual).toHaveProperty('expired', 2);
          expect(actual).toHaveProperty('nomatches', 1);
          expect(actual).toHaveProperty('missing', 4);
        });

    });
    it('should throw an error if the session is missing', () => {
      storageMock.retrieve.mockReturnValue(Promise.reject(new Error('Made to fail')));
      const fn = () => analytics.details('I am a session');
      expect(fn).rejects.toThrowErrorMatchingSnapshot();

    });
  });
});
