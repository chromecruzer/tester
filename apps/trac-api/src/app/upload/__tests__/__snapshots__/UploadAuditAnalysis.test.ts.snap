// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAuditAnalysis analytics should throw an error if the session is missing 1`] = `
"Session I am a session not found Error: Made to fail
    at Object.<anonymous> (/Users/mbletzin/Work/actalent/trac/apps/trac-api/src/app/upload/__tests__/UploadAuditAnalysis.test.ts:105:59)
    at Promise.then.completed (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/utils.js:333:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/utils.js:259:10)
    at _callCircusTest (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:277:40)
[90m    at processTicksAndRejections (node:internal/process/task_queues:96:5)[39m
    at _runTest (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:209:3)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:97:9)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:91:9)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:91:9)"
`;

exports[`UpdateAuditAnalysis details should return the correct filtered results 1`] = `
Object {
  "expired": Array [
    Object {
      "audit_match": "Expired",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid9",
    },
    Object {
      "audit_match": "Expired",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid10",
    },
  ],
  "location": Object {
    "name": "this is a location record",
    "scan_date": 2021-10-10T05:00:00.000Z,
  },
  "matches": Array [
    Object {
      "audit_match": "True Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid1",
    },
    Object {
      "audit_match": "True Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid2",
    },
    Object {
      "audit_match": "True Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid3",
    },
    Object {
      "audit_match": "True Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid4",
    },
    Object {
      "audit_match": "True Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid5",
    },
  ],
  "missing": Array [
    Object {
      "audit_match": "Missing",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid12",
    },
    Object {
      "audit_match": "Missing",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid13",
    },
    Object {
      "audit_match": "Missing",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid14",
    },
    Object {
      "audit_match": "Missing",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid15",
    },
  ],
  "moved": Array [
    Object {
      "audit_match": "Found In Other Location",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid6",
    },
    Object {
      "audit_match": "Found In Other Location",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid7",
    },
    Object {
      "audit_match": "Found In Other Location",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid8",
    },
  ],
  "nomatches": Array [
    Object {
      "audit_match": "No Match",
      "confirmed_date": null,
      "expire_date": null,
      "last_changed_date": null,
      "uuid": "uuid11",
    },
  ],
}
`;

exports[`UpdateAuditAnalysis details should throw an error if the session is missing 1`] = `
"Session I am a session not found Error: Made to fail
    at Object.<anonymous> (/Users/mbletzin/Work/actalent/trac/apps/trac-api/src/app/upload/__tests__/UploadAuditAnalysis.test.ts:87:57)
    at Promise.then.completed (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/utils.js:333:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/utils.js:259:10)
    at _callCircusTest (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:277:40)
[90m    at processTicksAndRejections (node:internal/process/task_queues:96:5)[39m
    at _runTest (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:209:3)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:97:9)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:91:9)
    at _runTestsForDescribeBlock (/Users/mbletzin/Work/actalent/trac/node_modules/[4mjest-circus[24m/build/run.js:91:9)"
`;
